src/Features/Core/Portable/Diagnostics/BuildOnlyDiagnosticsService.cs

```diff
@@ -37,12 +56,12 @@ private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
                 case WorkspaceChangeKind.SolutionCleared:
                 case WorkspaceChangeKind.SolutionReloaded:
                 case WorkspaceChangeKind.SolutionRemoved:
-                    ClearAllDiagnostics();
+                    _workQueue.AddWork(cancellationToken => ClearAllDiagnosticsAsync(cancellationToken));
                     break;
 
                 case WorkspaceChangeKind.ProjectReloaded:
                 case WorkspaceChangeKind.ProjectRemoved:
-                    ClearDiagnostics(e.OldSolution.GetProject(e.ProjectId));
+                    _workQueue.AddWork(cancellationToken => ClearDiagnosticsAsync(e.OldSolution.GetProject(e.ProjectId), cancellationToken));
```

Comment:
This is going to result in capturing `e` on every call to `OnWorkspaceChanged` (even the ones that don't use this method.

Reply:
godo call. will change.

Reply:
done.